### PR TITLE
Optimize the DCT GPU kernel.

### PR DESCRIPTION
**Pull request by banasraf**
#### Why we need this PR?
- Refactoring to improve the performance of the GPU DCT kernel. In the current state it's extremely slow for the transform done on the inner axis, so I handle this case separately. Also the existing CUDA kernel was slightly optimized.

#### What happened in this PR?
 - What solution was applied:
     The case with the transform done over the inner axis is handled with a separate CUDA kernel. The existing kernel was optimized by employing shared memory.
 - Affected modules and functionalities:
    GPU DCT kernel.
 - Key points relevant for the review:
     A new CUDA kernel and the changes in the old one.
 - Validation and testing:
     Existing tests still apply. I added a performance test.
 - Documentation (including examples):
     N/A

**JIRA TASK**: *DALI-1690*

### CODE REVIEW

**Comment by banasraf** [comment]
!build

**Comment by JanuszL** [comment]
What is the speed now, and what used to be before that optimization?

File: dali/kernels/signal/dct/dct_gpu.cu [hunk 1]
```diff
@@ -58,6 +67,54 @@ __global__ void ApplyDct(const typename Dct1DGpu<OutputType, InputType>::SampleD
   }
 }
 
+template <typename OutputType, typename InputType, bool HasLifter>
+__global__ void ApplyDctInner(const typename Dct1DGpu<OutputType, InputType>::SampleDesc *samples,
+                              const BlockSetupInner::BlockDesc *blocks,
+                              const float *lifter_coeffs) {
+  extern __shared__ char *shm[];
```
Comment (mzient) [inline dali/kernels/signal/dct/dct_gpu.cu:74]:
```suggestion
  extern __shared__ char shm[];
```

File: dali/kernels/signal/dct/dct_gpu.h [hunk 2]
```diff
@@ -33,6 +33,40 @@ namespace kernels {
 namespace signal {
 namespace dct {
 
+class BlockSetupInner {
+ public:
+  struct BlockDesc {
+    int64_t sample_idx;
+    int64_t frame_start;
+    int64_t frames_num;
```
Comment (mzient) [inline dali/kernels/signal/dct/dct_gpu.h:41]:
```suggestion
    int64_t num_frames;
```
or
```suggestion
    int64_t frame_count;
```

**Comment by banasraf** [comment]
!build

**Comment by banasraf** [comment]
@JanuszL 
> What is the speed now, and what used to be before that optimization?
For the planar layout it's ~550 GFLOPS -> ~630 GFLOPS and for interleaved it's ~30 GFLOPS -> ~315 GFLOPS

File: dali/kernels/signal/dct/dct_gpu.cu (hunk 1 above)
Comment (banasraf) [reply dali/kernels/signal/dct/dct_gpu.cu]: done

File: dali/kernels/signal/dct/dct_gpu.h (hunk 2 above)
Comment (banasraf) [reply dali/kernels/signal/dct/dct_gpu.h]: done

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/test_operator_decoder.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def define_graph(self):
 test_data_root = get_dali_extra_path()
 good_path = 'db/single'
 missnamed_path = 'db/single/missnamed'
-test_good_path = {'jpeg', 'mixed', 'png', 'tiff', 'pnm', 'bmp', 'jpeg2k'}
+test_good_path = {'jpeg2k'}
```
Comment (JanuszL) [inline dali/test/python/test_operator_decoder.py:43]: ???

Comment (banasraf) [reply dali/test/python/test_operator_decoder.py:43]: eeeh, good catch. I've committed too much

Comment (banasraf) [reply dali/test/python/test_operator_decoder.py]: done

**Review by JanuszL — APPROVED** [review]

**Comment by banasraf** [comment]
!build

**Review by mzient — APPROVED** [review]

**Pull request merged**